File: articles/calculus-for-systems-modeling/articles/inverse-functions-and-system-interpretation/julia/inverse_interpretation_audit.jl
forward_model(x) = log1p(x)
forward_derivative(x) = 1.0 / (1.0 + x)
inverse_model(y) = exp(y) - 1.0

println("target_output,recovered_input,forward_check,residual,forward_derivative,inverse_sensitivity,domain_valid")
for y in [0.0, 0.5, 1.0, 1.5, 2.0]
    x = inverse_model(y)
    ycheck = forward_model(x)
    residual = ycheck - y
    deriv = forward_derivative(x)
    invsens = 1.0 / deriv
    domainvalid = x > -1.0
    println("$y,$x,$ycheck,$residual,$deriv,$invsens,$domainvalid")
end

File: articles/calculus-for-systems-modeling/articles/inverse-functions-and-system-interpretation/cpp/inverse_interpretation_audit.cpp
#include <cmath>
#include <iomanip>
#include <iostream>
#include <vector>

double forward_model(double x){ return std::log1p(x); }
double forward_derivative(double x){ return 1.0 / (1.0 + x); }
double inverse_model(double y){ return std::exp(y) - 1.0; }

int main(){
  std::vector<double> ys={0.0,0.5,1.0,1.5,2.0};
  std::cout<<std::fixed<<std::setprecision(12);
  std::cout<<"target_output,recovered_input,forward_check,residual,forward_derivative,inverse_sensitivity,domain_valid\n";
  for(double y:ys){
    double x=inverse_model(y), ycheck=forward_model(x), residual=ycheck-y;
    double derivative=forward_derivative(x), invsens=1.0/derivative;
    bool domain_valid=x > -1.0;
    std::cout<<y<<","<<x<<","<<ycheck<<","<<residual<<","<<derivative<<","<<invsens<<","<<domain_valid<<"\n";
  }
}
